WordPress/gutenberg
# Release Workflow: Allow triggering manually

## Description
Allow manual triggering of the workflow to create a new release, per [this discussion](https://github.com/WordPress/gutenberg/pull/27488#issuecomment-757946026).

The end goal is for the release manager to run the entire release process from GitHub, and to minimize manual interaction.
- :hand: Release manager starts release process manually.
- :robot: GitHub workflow bumps version numbers in plugin files, builds Gutenberg plugin, creates release draft, and prefills it with release notes based on changelog script output.
- :hand: Release manager edits release draft, and eventually publishes it. This creates the tag.
- :robot: (Other) GitHub workflow tags prepends release notes to `changelog.txt`, uploads Gutenberg plugin and updated `changelog.txt` to WP.org plugin repo.

Functionally, this is as close as possible to the previous workflow. The major difference is that `changelog.txt` is only updated _after_ tagging. This is a result of using the GitHub release draft UI to edit and revise release notes. It's not possible to include the result of those changes _before_ tagging.

## Demo

https://youtu.be/TnSgJd3zpJY

## How has this been tested?

See here for an example release created by this action: https://github.com/ockham/gutenberg/releases/tag/v10.2.0

Corresponding workflow run: https://github.com/ockham/gutenberg/actions/runs/654286491

Or use the following instructions to test for yourself:

- Make sure you have your own fork of the Gutenberg repo, and that you've added it locally as a remote (E.g. `git remote add ockham git@github.com:ockham/gutenberg.git` if your GH username is `ockham` :slightly_smiling_face: ).
- Check out this PR's branch, and push the contents to your fork's `trunk` branch (you can reset later):
```
git push ockham update/trigger-release-workflow-manually:trunk --force
```
- Do the same for the current release branch, e.g.
```
git push ockham update/trigger-release-workflow-manually:release/10.2 --force
```
- Go to the 'Actions' tab of your fork's GH page (e.g. https://github.com/ockham/gutenberg/actions).
- In the left sidebar (list of available actions), locate the action whose description reads "Build Gutenberg and upload to Draft Release", and click on it.
- Locate the blue "This workflow has a workflow_dispatch event trigger." header, and click the "Run workflow" button next to it.
- Keep the `trunk` branch as the base branch, and specify either `stable` or `rc` to release create a stable or release candidate version of the plugin.
- It should take less than 5 minutes to build the plugin zip. You can watch that page as the action progresses.
- Eventually, a release draft will be created, and the plugin zip will be attached. You'll find it under 'Releases' (e.g. https://github.com/ockham/gutenberg/releases)
- Download the release.
- Verify that the plugin version has indeed been updated (in `gutenberg.php`, `package.json`, etc), and that the Changelog contains the release notes for the release.
- Verify that the newly created _tag_ for the release _also_ has the version bump, _especially_ when creating an RC!
- Publish the release, and verify that the release notes are correctly used to update `changelog.txt` on both `trunk`, and the release branch.

(I've used my Gutenberg fork as an example, you can see the relevant action and release drafts from previous runs there.)

## Follow-up:
- [ ] Remove old release script
- [ ] Modify svn logic to allow uploading point releases for previous version series. (Use the correct changelog, and to only set the "Stable Version" thing if we are updating the stable version.)
